containerd-shim-v2/create.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2014,2015,2016 Docker, Inc.
+// Copyright (c) 2017 Intel Corporation
+// Copyright (c) 2018 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package containerdshim
+
+import (
+	"context"
+	"fmt"
+
+	vc "github.com/kata-containers/runtime/virtcontainers"
+	"github.com/kata-containers/runtime/virtcontainers/pkg/oci"
+
+	taskAPI "github.com/containerd/containerd/runtime/v2/task"
+
+	"github.com/kata-containers/runtime/pkg/katautils"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func create(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest, netns string,
+	runtimeConfig *oci.RuntimeConfig) (*container, error) {
+
+	detach := !r.Terminal
+
+	// Checks the MUST and MUST NOT from OCI runtime specification
+	bundlePath, err := validCreateParams(r.ID, r.Bundle)
+	if err != nil {
+		return nil, err
+	}
+
+	ociSpec, err := oci.ParseConfigJSON(bundlePath)
+	if err != nil {
+		return nil, err
+	}
+
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return nil, err
+	}
+
+	//In the sandbox, the containers will only
+	//use the mnt space to separate the rootfs,
+	//and to share the other namesapces with host
+	//in the sandbox, thus remove those namespaces
+	//from ocispec except networkNamespace, since
+	//it has been ignored by kata-agent in sandbox.
+
+	for _, ns := range []specs.LinuxNamespaceType{
+		specs.UserNamespace,
+		specs.UTSNamespace,
+		specs.IPCNamespace,
+		specs.PIDNamespace,
+		specs.CgroupNamespace,
+	} {
+		removeNameSpace(&ociSpec, ns)
```

Comment:
Yes, it must be removed. For kubernetes,  all of the containers in a pod will share the same namespaces except mnt namespace,  thus the container's ocispec contains the pod's namespace info such as "/proc/<pid>/ns/pid", but the "pid" is wrong, shimv2 returned a fake pid to containerd. Thus we must remove them otherwise the kata agent will complained of "cannot find the pid" error.

Reply:
I see. Then I think it's not the containerd-shim-v2 specific problem, kata-runtime cli must have same problem. I think it's better to do some converting(to right pid) or removing in somewhere as `virtcontainers/kata-agent.go`, as a common handler

Reply:
I had double checked that kata-agent.go had dealt  the  namespaces.  Thus here it shouldn't deal with those namespaces, but there is a bug in dealing with Pid namespace in virtcontainers/kata-agent.go and there is also a bug in kata agent for dealing with "sharedPidNs", which will complained error as below: 
  Could not run process: container_linux.go:348: starting container process caused "process_linux.go:301: running exec setns process for init caused \"exit status 40\"": unknownill 

Thus here I will just remove the pid namespace and once I fixed the above two issues then I will delete these "remove" lines codes.

Reply:
It's better to file an issue tracking this, in case we leave it behind

Reply:
yes, had opened an issue as https://github.com/kata-containers/runtime/issues/930